apps/managedidentity/managedidentity_test.go

```diff
@@ -4,43 +4,401 @@ package managedidentity
+				GrantedScopes: accesstokens.Scopes{Slice: tokenScope},
+				TokenType:     "TokenType",
+			}
+			if result.AccessToken != expected.AccessToken {
+				t.Fatalf(`unexpected access token "%s"`, result.AccessToken)
+			}
+		})
+	}
+}
+
+func TestCreatingIMDSClient(t *testing.T) {
+	tests := []struct {
+		name    string
+		id      ID
+		wantErr bool
+	}{
+		{
+			name: "System Assigned",
+			id:   SystemAssigned(),
+		},
+		{
+			name: "Client ID",
+			id:   ClientID("test-client-id"),
+		},
+		{
+			name: "Resource ID",
+			id:   ResourceID("test-resource-id"),
+		},
+		{
+			name: "Object ID",
+			id:   ObjectID("test-object-id"),
+		},
+		{
+			name:    "Empty Client ID",
+			id:      ClientID(""),
+			wantErr: true,
+		},
+		{
+			name:    "Empty Resource ID",
+			id:      ResourceID(""),
+			wantErr: true,
+		},
+		{
+			name:    "Empty Object ID",
+			id:      ObjectID(""),
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client, err := New(tt.id)
+			if tt.wantErr {
+				if err == nil {
+					t.Fatal("client New() should return a error but did not.")
+				}
+				return
+			}
+			if err != nil {
+				t.Fatal("client New() error while creating client")
+			} else {
+				if client.miType.value() != tt.id.value() {
+					t.Fatal("client New() did not assign a correct value to type.")
+				}
+			}
+		})
+	}
+
+}
+func TestCreateIMDSAuthRequest(t *testing.T) {
+	tests := []struct {
+		name     string
+		id       ID
+		resource string
+		claims   string
+		wantErr  bool
+	}{
+		{
+			name:     "System Assigned",
+			id:       SystemAssigned(),
+			resource: "https://management.azure.com",
+		},
+		{
+			name:     "System Assigned",
+			id:       SystemAssigned(),
+			resource: "https://management.azure.com/.default",
+		},
+		{
+			name:     "Client ID",
+			id:       ClientID("test-client-id"),
+			resource: "https://storage.azure.com",
+		},
+		{
+			name:     "Resource ID",
+			id:       ResourceID("test-resource-id"),
+			resource: "https://vault.azure.net",
+		},
+		{
+			name:     "Object ID",
+			id:       ObjectID("test-object-id"),
+			resource: "https://graph.microsoft.com",
+		},
+		{
+			name:     "With Claims",
+			id:       SystemAssigned(),
+			resource: "https://management.azure.com",
+			claims:   "test-claims",
+		},
 	}
 
-	_, err = client.AcquireToken(context.Background(), "scope", WithClaims("claim"))
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			req, err := createIMDSAuthRequest(context.Background(), tt.id, tt.resource, tt.claims)
+			if tt.wantErr {
+				if err == nil {
+					t.Fatal(err)
+				}
+				return
+			}
+			if req == nil {
+				t.Fatal("createIMDSAuthRequest() returned nil request")
+				return
+			}
+			if req.Method != http.MethodGet {
+				t.Fatal("createIMDSAuthRequest() method is not GET")
+			}
+			if !strings.HasPrefix(req.URL.String(), imdsEndpoint) {
+				t.Fatal("createIMDSAuthRequest() URL is not matched.")
```

Comment:
A test failure like this is easier to debug when the message includes want/got values
```suggestion
			if got := req.URL.String(); !strings.HasPrefix(got, imdsEndpoint) {
				t.Fatalf("wanted %q, got %q", imdsEndpoint, got)
```